Adin1110 add support

tachicialex opened the pull request:
The ADIN1110 is a low power single port 10BASE-T1L MAC-PHY
designed for industrial Ethernet applications. It integrates
an Ethernet PHY core with a MAC and all the associated analog
circuitry, input and output clock buffering.
    
ADIN1110 MAC-PHY encapsulates the ADIN1100 PHY. The PHY registers
can be accessed through the the MDIO MAC registers. We
are registering an MDIO bus with custom read/write in order
to let the PHY to be discovered by the PAL. This will let
the ADIN1100 Linux driver to probe and take control of
the PHY.

b16339a7703a - Kconfig.adi: select ADIN1110 for build
c952e0d8a1ec - dt-bindings: net: adin1110: Add docs
3c3e4537828c - net: adi: adin1110: Add initial support
c581fa48199a - net: phy: adin1100: Add ADIN1110 PHY_ID

While not currently upstreamed, it is needed in multiple projects.
I plan to upstream this one immediately after adin1100 and
10Base-T1L get accepted.

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+	u32 status1;
+
+	while (1) {
+		adin1110_read_reg(priv, ADIN1110_STATUS1, &status1);
```
no error handling... I guess we break right way if an error is returned?

nunojsa replied:
this is still open...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+		if (status1 & ADIN1110_RX_RDY)
+			adin1110_read_fifo(priv);
+		else
+			break;
```
Maybe this would look neater?

```
if (!(status1 & ADIN1110_RX_RDY))
     break;

adin1110_read_fifo(priv);
```

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+
+	rxpkt = skb_put(rxb, frame_size_no_fcs);
+	memcpy(rxpkt, &priv->data[header_len + ADIN1110_FRAME_HEADER_LEN],
+	       frame_size_no_fcs);
```
[skb_copy_to_linear_data()](https://elixir.bootlin.com/linux/latest/C/ident/skb_copy_to_linear_data)?

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+
+	frame_size_no_fcs = frame_size - ADIN1110_FRAME_HEADER_LEN - ADIN1110_FEC_LEN;
+
+	rxb = netdev_alloc_skb(priv->netdev, frame_size_no_fcs);
```
curious... How and when is this data freed?

tachicialex replied:
After reading this I assumed they are freed at socket level (tcp/udp stuff): https://patchwork.kernel.org/project/netdevbpf/patch/20211115190249.3936899-18-eric.dumazet@gmail.com/

nunojsa reviewed (commented):
Looks good... Some general comments from my side as I'm not that familiar with phy framework... Also please fix the minor issue reported by checkpatch

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+	fwnode = dev_fwnode(dev);
+
+	/* use of CRC on control and data transactions is pin dependent */
+	if (fwnode_property_present(fwnode, "adi,spi-crc")) {
```
here, probably it's actually better to have `priv->append_crc = device_property_read_bool()`

tachicialex commented:
Changelog v1 -> v2:
  - in: adin1110_read_fifo(): use skb_copy_to_linear_data() instead of memcpy()
  - added error checking in adin1110_read_mdio_acc()
  - in adin1110_register_mdiobus() replaced mdiobus_register() with devm_mdiobus_register()
  - in adin1110_write_mac_address() write given mask instead of always FF FF FF FF FF FF
  - in adin1110_get_drvinfo() replaced strlcpy with strscpy
  - removed locking from adin1110_check_spi()
  - in adin1110_probe() use device_property_read_bool() instead of device_property_read_bool()
  - in adin1110_probe() use device_get_mac_address() instead of fwnode_get_mac_address()
  - removed adin1110_remove() from struct spi_driver;
  - added error checking for phy_connect()
  - added devm action in order to call phy_disconnect()

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,972 @@
+	int ret;
+
+	memset(mask, 0xFF, ETH_ALEN);
+
```
nit: below you don't have new line and here you have... maybe be consistent (personally I prefer how you have without new line)

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,974 @@
+
+	if (accept_multicast) {
+		mask[0] = BIT(1);
+		mac[0] = BIT(1);
```
I will assume this is on purpose...

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,972 @@
+	ret = adin1110_read_reg(priv, ADIN1110_TX_SPACE, &val);
+	if (ret < 0) {
+		mutex_unlock(&priv->lock);
+
```
nit: no new line

nunojsa commented on drivers/net/ethernet/adi/adin1110.c:
```diff
@@ -0,0 +1,972 @@
+	 * The FEC will be added by the MAC internally.
+	 */
+	if (txb->len + ADIN1110_FEC_LEN < 64)
+		padding = 64 - (txb->len + ADIN1110_FEC_LEN);
```
nit: new line

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -0,0 +1,95 @@
+  interrupts:
+    maxItems: 1
+
+  phy@0:
```
If mandatory this one needs to be on the "required:" list

nunojsa reviewed (commented):
still some open stuff

nunojsa commented on Documentation/devicetree/bindings/net/adi,adin1110.yaml:
```diff
@@ -0,0 +1,96 @@
+  - compatible
+  - reg
+  - interrupts
+  - phy
```
Did you validated your binding? I think this should by phy@0 or the the validations might fail... Not sure though

tachicialex replied:
Just tried phy and phy@0. Does not seem to complain about neither of them.

nunojsa reviewed (approved):
It would be nice to make sure the binding is actually being properly validated (see my comment below)
